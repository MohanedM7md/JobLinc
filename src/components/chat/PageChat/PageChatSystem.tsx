import { useCallback, useEffect, useState } from "react";
import ChatCardsList from "../ChatCardsList";
import PageMessageWindow from "./PageChatWindow";
import useChatId from "@context/ChatIdProvider";
import connectToChat, {
  disconnectChatSocket,
  onConnect,
} from "@services/api/ChatSocket";
import store from "@store/store";
import { AnimatePresence, motion } from "framer-motion";
import { Menu, X, Users2 } from "lucide-react";
import SearchBar from "@chatComponent/UI/SearchBar";
import NetWorksChatList from "@chatComponent/NetWorksChatList";
import useNetworkUserId from "@context/NetworkUserIdProvider";
import ConnectionsSidebar from "@chatComponent/PageChat/ConnectionsSidebar";
import ConfirmationModal from "@chatComponent/UI/ConfirmationModal";
import { BlockMessaging } from "@services/api/chatServices";
import ToggleSwitch from "../UI/ToggleSwitch";
// Mock API function to fetch messaging status (replace with actual implementation)
const getMessagingStatus = async (): Promise<boolean> => {
  try {
    return false; // Default mock value
  } catch (error) {
    console.error("Failed to fetch messaging status:", error);
    return false; // Default to unblocked on error
  }
};

function PageChatSystem() {
  const { setChatId } = useChatId();
  const { setUsersId } = useNetworkUserId();
  const [isFocused, setIsFocused] = useState<boolean>(false);
  const [isSidebarOpen, setIsSidebarOpen] = useState<boolean>(false);
  const [isChatSidebarOpen, setIsChatSidebarOpen] = useState<boolean>(true);
  const [searchTerm, setSearchTerm] = useState<string>("");
  const [opnedChatName, setOpnedChatName] = useState<string>("");
  const [isConnected, setIsConnected] = useState<boolean>(false);
  const [isMobileView, setIsMobileView] = useState<boolean>(false);

  const [showBlockConfirmation, setShowBlockConfirmation] = useState(false);
  const [isBlocked, setIsBlocked] = useState<boolean>(
    !store.getState().user.allowMessages!,
  );
  const [pendingBlockState, setPendingBlockState] = useState<boolean | null>(
    null,
  );

  const toggleBlockMessaging = () => {
    setPendingBlockState(isBlocked);
    setShowBlockConfirmation(true);
  };
  const handleBlockUser = async () => {
    if (pendingBlockState === null) return;
    try {
      console.log(
        "send to the back end the wanted status: ",
        pendingBlockState,
      );
      setShowBlockConfirmation(false);
      await BlockMessaging(pendingBlockState);
      setIsBlocked(!pendingBlockState);
      setPendingBlockState(null);
    } catch (error) {
      console.error(
        pendingBlockState
          ? "Failed to block messaging:"
          : "Failed to unblock messaging:",
        error,
      );
      setShowBlockConfirmation(false);
    }
  };

  const handleCancelToggle = () => {
    setShowBlockConfirmation(false);
    setPendingBlockState(null);
  };

  useEffect(() => {
    const checkMobileView = () => {
      setIsMobileView(window.innerWidth < 768);

      if (window.innerWidth < 768) {
        setIsChatSidebarOpen(false);
      } else {
        setIsChatSidebarOpen(true);
      }
    };

    checkMobileView();
    window.addEventListener("resize", checkMobileView);
    return () => window.removeEventListener("resize", checkMobileView);
  }, []);

  const onFocusedToggler = () => {
    setTimeout(() => {
      setIsFocused(!isFocused);
    }, 100);
  };

  const handleSearchChange = useCallback(
    (value: string) => setSearchTerm(value),
    [],
  );

  const handleConversationClick = useCallback(
    (chatId: string, chatName: string) => {
      setUsersId([]);
      setChatId(chatId);
      setOpnedChatName(chatName);

      if (isMobileView) {
        setIsChatSidebarOpen(false);
      }
    },
    [isMobileView],
  );

  const handleNetWorkUserClick = useCallback(
    (userId: string, chatName: string) => {
      setUsersId([userId]);
      setOpnedChatName(chatName);
      setChatId("");

      if (isMobileView) {
        setIsChatSidebarOpen(false);
      }
    },
    [isMobileView],
  );

  useEffect(() => {
    const initializeChat = async () => {
      connectToChat();
      onConnect(setIsConnected);
    };

    initializeChat();
    return () => disconnectChatSocket();
  }, []);

  return (
    <div className="h-full flex flex-col">
      <header className="flex items-center justify-between bg-gray-100 p-2 sm:p-3 rounded-lg shadow-md">
        <div className="flex items-center gap-2">
          {isMobileView && (
            <button
              onClick={() => setIsChatSidebarOpen(!isChatSidebarOpen)}
              className="p-1 sm:p-2 rounded-full hover:bg-gray-200 transition"
              aria-label="Toggle chat sidebar"
            >
              {isChatSidebarOpen ? (
                <X className="w-4 h-4 sm:w-5 sm:h-5" />
              ) : (
                <Menu className="w-4 h-4 sm:w-5 sm:h-5" />
              )}
            </button>
          )}
          <h2 className="hidden md:block font-semibold text-lg text-gray-800">
            Messaging
          </h2>
          {isMobileView && !isChatSidebarOpen && (
            <h2 className="font-semibold text-base text-gray-800 truncate max-w-[100px] sm:max-w-[150px]">
              {opnedChatName || "Messaging"}
            </h2>
          )}
        </div>

        <div className="flex items-center gap-1 sm:gap-2 w-full justify-end">
          <SearchBar
            FocusToggler={onFocusedToggler}
            onChange={handleSearchChange}
            className="w-32 sm:w-48 md:w-64"
          />
          <button
            onClick={() => setIsSidebarOpen(true)}
            className="p-1 sm:p-2 rounded-full hover:bg-gray-200 transition"
          >
            <Users2 className="w-4 h-4 sm:w-5 sm:h-5" />
          </button>
          {isSidebarOpen && (
            <ConnectionsSidebar
              isOpen={isSidebarOpen}
              onClose={() => setIsSidebarOpen(false)}
            />
          )}
          <div className="relative">
            <ToggleSwitch
              checked={isBlocked}
              onChange={toggleBlockMessaging}
              label="Block Messaging"
              ariaLabel="Toggle block messaging"
              className=""
            />
          </div>

          <ConfirmationModal
            isOpen={showBlockConfirmation}
            onClose={handleCancelToggle}
            onConfirm={handleBlockUser}
            title={pendingBlockState ? "Unblock Messaging" : "Block Messaging"}
            confirmText="Confirm"
            isDangerous={pendingBlockState || false}
          >
            <p className="text-gray-600 font-bold text-2xl">
              {pendingBlockState
                ? "Finally you decided to stop being karzmaðŸ˜€"
                : "NOOOOOO please. No one will talk to you again ðŸ˜­"}
            </p>
          </ConfirmationModal>
        </div>
      </header>

      {isConnected && (
        <div className="flex flex-1 overflow-hidden">
          <AnimatePresence mode="wait" initial={false}>
            {isChatSidebarOpen && (
              <motion.div
                initial={
                  isMobileView ? { x: -300, opacity: 0 } : { opacity: 1 }
                }
                animate={{ x: 0, opacity: 1 }}
                exit={isMobileView ? { x: -300, opacity: 0 } : { opacity: 0 }}
                transition={{ duration: 0.3 }}
                className={`
                  ${
                    isMobileView
                      ? "absolute z-10 h-[calc(100%-70px)] bg-white"
                      : "relative"
                  } 
                  w-full md:w-1/3 border-r bg-charcoalWhite border-gray-300 p-4 overflow-y-scroll
                `}
              >
                {isFocused ? (
                  <NetWorksChatList
                    onCardClick={handleNetWorkUserClick}
                    filter={searchTerm}
                  />
                ) : (
                  <ChatCardsList
                    onCardClick={handleConversationClick}
                    className="overflow-y-scrolls"
                  />
                )}
              </motion.div>
            )}
          </AnimatePresence>

          <PageMessageWindow chatName={opnedChatName} />
        </div>
      )}
    </div>
  );
}

export default PageChatSystem;
