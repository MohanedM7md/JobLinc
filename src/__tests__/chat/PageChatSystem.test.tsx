import { render, screen, act } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import PageChatSystem from "@chatComponent/PageChat/PageChatSystem";

vi.mock("@services/api/ChatSocket", () => ({
  connectToChat: vi.fn(),
  disconnectChatSocket: vi.fn(),
  onConnect: vi.fn((callback) => {
    setTimeout(() => callback(true), 0); // Automatically call the callback with true
    return vi.fn();
  }),
}));

vi.mock("@services/api/chatServices", () => ({
  BlockMessaging: vi.fn().mockResolvedValue(true),
}));

// Mock context providers
vi.mock("@context/ChatIdProvider", () => ({
  default: () => ({
    setChatId: vi.fn(),
  }),
}));

vi.mock("@context/NetworkUserIdProvider", () => ({
  default: () => ({
    setUsersId: vi.fn(),
  }),
}));

vi.mock("@chatComponent/ChatCardsList", () => ({
  __esModule: true,
  default: () => <div data-testid="chat-list">Mocked ChatCardsList</div>,
}));

vi.mock("@chatComponent/NetWorksChatList", () => ({
  __esModule: true,
  default: () => <div>Mocked NetWorksChatList</div>,
}));

vi.mock("@chatComponent/PageChat/PageChatWindow", () => ({
  default: (props) => (
    <div data-testid="page-message-window" data-chatname={props.chatName}>
      Message Window
    </div>
  ),
}));

vi.mock("@chatComponent/PageChat/ConnectionsSidebar", () => ({
  default: (props) =>
    props.isOpen ? (
      <div data-testid="connections-sidebar">
        <button data-testid="close-sidebar" onClick={props.onClose}>
          Close
        </button>
      </div>
    ) : null,
}));

vi.mock("@chatComponent/UI/SearchBar", () => ({
  default: (props) => (
    <div data-testid="search-bar" className={props.className}>
      <input
        data-testid="search-input"
        onChange={(e) => props.onChange(e.target.value)}
        onFocus={props.FocusToggler}
        onBlur={props.FocusToggler}
      />
    </div>
  ),
}));

describe("PageChatSystem", () => {
  it("renders ChatCardsList with mocked useState", async () => {
    vi.mock("react", async () => {
      let call = 0;
      const actual = await vi.importActual<typeof import("react")>("react");
      return {
        ...actual,
        useState: vi.fn(() => {
          const mockedStates = [
            [false, vi.fn()], // isFocused
            [true, vi.fn()], // isSidebarOpen (opened)
            [true, vi.fn()], // isChatSidebarOpen
            ["", vi.fn()], // searchTerm
            ["", vi.fn()], // opnedChatName
            [true, vi.fn()], // isConnected
            [false, vi.fn()], // isMobileView
            [false, vi.fn()], // menuOpen
            [false, vi.fn()], // showBlockConfirmation
          ];
          const state = mockedStates[call % mockedStates.length];
          call++;
          return state;
        }),
      };
    });
    render(<PageChatSystem />);
    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 10));
    });
    expect(screen.getByText("Messaging")).toBeInTheDocument();
    expect(screen.getByTestId("chat-list")).toBeInTheDocument();
    expect(screen.getByTestId("page-message-window")).toBeInTheDocument();
  });
});
