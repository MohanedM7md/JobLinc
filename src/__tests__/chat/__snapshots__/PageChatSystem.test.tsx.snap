// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PageChatSystem  > Should RenderChatSystem:  1`] = `
<div>
  <div
    class="flex items-center justify-between bg-gray-100 p-3 rounded-lg shadow-md"
  >
    <h2
      class="font-semibold text-lg text-gray-800"
    >
      Messaging
    </h2>
    <div
      class="flex items-center gap-3"
    >
      <div
        class="h-10 flex items-center border border-gray-300 rounded-lg px-2 cursor-pointer bg-white 
        hover:border-gray-400"
      >
        <input
          class="border-none outline-none text-sm caret-crimsonRed flex-grow"
          name="text"
          placeholder="Search..."
          type="text"
          value=""
        />
        <svg
          class="lucide lucide-search w-4 h-4 text-gray-500"
          data-testid="test-search-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <path
            d="m21 21-4.3-4.3"
          />
        </svg>
      </div>
      <button
        class="p-2 rounded-full hover:bg-gray-200"
      >
        <svg
          class="lucide lucide-ellipsis-vertical w-6 h-6 text-gray-600"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="1"
          />
          <circle
            cx="12"
            cy="5"
            r="1"
          />
          <circle
            cx="12"
            cy="19"
            r="1"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    class="flex h-screen w-full"
  >
    <div
      class="w-1/3 border-r bg-charcoalWhite border-gray-300 p-4 overflow-y-auto"
    />
    <div
      class="flex-grow flex flex-col justify-end w-2/3 flex flex-col"
      data-testid="test-PageWindow"
    >
      <div
        class="flex-grow flex items-center justify-center"
      >
        <h2
          class="text-gray-500"
        >
          Choose a chat to start messaging
        </h2>
      </div>
    </div>
  </div>
</div>
`;
